### ISSUE TITLE
Restoring from background sometimes gets stuck with 'Connection lost. Reconnecting…' for 10 seconds

### ISSUE THREAD

**Issue by mbaran5** [home-assistant/iOS#592]
**iOS device model, version and app version**

Model Name: iPhone 11 Pro
Software Version: 13.5 
App version: 2019.1

**Home Assistant Core Version**
0.110.1 w/temporary webhook.py fix for sim2 errors 

**Describe the bug**
I receive many "Disconnected: Did not receive auth message within 10 seconds" errors from the iOS mobile app. 

```
2020-05-28 10:03:08.638 [Verbose] [main] [RegionManager.swift:243] locationManager(_:didUpdateLocations:) > didUpdateLocations [<+33.00000000,-118.00000000> +/- 65.00m (speed -1.00 mps / course -1.00) @ 5/28/20, 12:29:47 AM Pacific Daylight Time]
2020-05-28 10:03:08.639 [Warning] [main] [RegionManager.swift:268] locationManager(_:didUpdateLocations:) > Location is older than threshold.
2020-05-28 10:03:08.732 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:08.732 [Error] [main] [TokenManager.swift:195] currentToken > Token is expired by 6405 seconds: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiIyYjZmOTkzYTg2YTc0ODhhYmViYjhlYjdiZDkwOWRhYyIsImlhdCI6MTU5MDY3NzE4MiwiZXhwIjoxNTkwNjc4OTgyfQ.KNE2ckvmVrx8myJfkAuNRc1zKfaSdsAAYR6dTCeWMh4
2020-05-28 10:03:08.885 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.035 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.035 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:11.675 [Verbose] [main] [HAAPI.swift:506] GetAndSendLocation(trigger:) > getAndSendLocation called via Manual
2020-05-28 10:03:11.676 [Verbose] [main] [OneShotLocationManager.swift:36] init(onLocation:) > LocationManager: Got location stored on manager <+33.00000000,-118.00000000> +/- 65.00m (speed -1.00 mps / course -1.00) @ 5/28/20, 10:03:08 AM Pacific Daylight Time
2020-05-28 10:03:11.683 [Info] [main] [HAAPI.swift:477] SubmitLocation(updateType:location:zone:) > Location update payload: ["altitude": 23.2812557220459, "vertical_accuracy": 10.0, "gps": [33.838134765625, -118.33911486011463], "battery": 99, "gps_accuracy": 65.0]
```

Others on the HA forum are experiencing similar issues 

- https://github.com/home-assistant/core/issues/34711

**To Reproduce**
Normal app usage after a short period of time of app being in background. 

**Expected behavior**
App not to throw errors in HA logs

**Screenshots**
<img width="589" alt="image" src="https://user-images.githubusercontent.com/583865/82833146-168cad80-9e72-11ea-8fc6-bdc3c27c4c71.png">

**Comment by TomBrien** [comment]
Do you see the same behaviour when using the official `mobile_app` integration that ships with Home Assistant

**Comment by mbaran5** [comment]
I'm using default_config: in configuration.yaml so I don't have a mobile_app: in there. 

```
# Configure a default setup of Home Assistant (frontend, api, etc)
default_config:
```

**Comment by TomBrien** [comment]
`default_config` calls `mobile_app`. As I understand it the fix involves add a custom component with the same name.  
> App version: 2019.1
> 
> 
> 
> **Home Assistant Core Version**
> 
> 0.110.1 w/temporary webhook.py fix for sim2 errors 
>

**Comment by mbaran5** [comment]
I'll pull that out and give it a shot. I'm not sure how it will reproduce as the mobile app has limited functionality with iOS 13.4+ but will see if it throws similar errors.

**Comment by mbaran5** [comment]
The issue occurs regardless of using mobile_app as a part of 0.110.2 or using the modified webhook.py as a part of the temporary fix for sim2. I may have narrowed it down to a camera_stream as a part of a picture_entity or button_card. I'll keep tracking and try to update here with my findings.

**Comment by mbaran5** [comment]
Ok, here's what I can put together so far. After some period of time (usually > 4 hours) my phone will go into a state where it's not able to communicate with HA. When I open the app, three things will happen.

1. I will get a connection lost notice in the HA iOS app
2. I will get the 10 second auth message in HA (and sometimes an auth failed for ip 10.0.x.x) 
3. I will notice that my device is showing GPS: 2500. 

The interesting thing about issue 3 is, it's related to my other open issue below. 
- https://github.com/home-assistant/iOS/issues/589 

That said, I don't know what causes this authentication timeout to occur in the first place. If I log into HA from my desktop browser, I will see my phone in one of two states, either with good GPS accuracy or bad. If bad (GPS: 2500) I will know that when I open my app I'll see it with the connection lost, reconnection message. That takes over 10 seconds and throws the error related to the topic of this issue. 

The only thing I can think of, is that it may be related to these two issues. I do have a camera in one of my lovelace dashboards that refreshes periodically. 
- https://github.com/home-assistant/core/issues/23055
- https://github.com/home-assistant/frontend/issues/3379

**Comment by TomBrien** [comment]
I have a suspicion this is actually a duplicate of #426 I haven't seen the credential bug in ages but I do see the Connection lost error. We believed this was an issue with the access token not getting refreshed properly. A test is to leave the app open put airplane mode on for 15-30 minutes (including with wi-fi off i.e. no data at all), turn airplane mode off again and open the app, this should trigger the behaviour

**Comment by mbaran5** [comment]
I'll give that a shot. Just to be clear the steps are

1. Open App
2. Toggle Airplane mode (complete data shut off)
3. Leave app open? or can I just turn off phone
4. Wait 30 minutes
5. Toggle data back on 
6. Open app and determine if it refreshes on it's own?

**Comment by TomBrien** [comment]
I think leave app open in background was the approach I used

**Comment by mbaran5** [comment]
I could not reproduce #426. I opened the app to make sure it was functioning, put the app in the background, enabled full airplane/no data mode for 35 minutes, turned off airplane mode and then opened the app. It said reconnecting but for just maybe 1-2 seconds tops and went back to working fine. No errors logged on the HA side, no 10 second timeout triggered. 

Is there a higher logging level I can enable to see where it's getting caught.

**Comment by TomBrien** [comment]
You didn't see the behaviour you described above after coming out of airplane mode? My gut tells me iOS is throttling the app here

**Comment by mbaran5** [comment]
Correct. 

It briefly said "Connection Lost. Reconnecting..." but it was short, maybe 1-2 seconds. It triggered no errors in either the app or HA. My GPS also remained correct and did not back out to the 2500 value it does when I normally see this issue. 

I've also never had it forget the login, just more of a long (10s+) "Connection Lost" message which usually indicates that it's timed out, it'll throw an HA log error, and my GPS will go wonky.

**Comment by TomBrien** [comment]
Might be worth checking the event log in the app

**Comment by mbaran5** [comment]
From the logs, shortly before 10:12AM (when I posted this morning) here's what happened
```
2020-05-28 10:03:08.638 [Verbose] [main] [RegionManager.swift:243] locationManager(_:didUpdateLocations:) > didUpdateLocations [<+33.83803790,-118.33891126> +/- 65.00m (speed -1.00 mps / course -1.00) @ 5/28/20, 12:29:47 AM Pacific Daylight Time]
2020-05-28 10:03:08.639 [Warning] [main] [RegionManager.swift:268] locationManager(_:didUpdateLocations:) > Location is older than threshold.
2020-05-28 10:03:08.732 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:08.732 [Error] [main] [TokenManager.swift:195] currentToken > Token is expired by 6405 seconds: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiIyYjZmOTkzYTg2YTc0ODhhYmViYjhlYjdiZDkwOWRhYyIsImlhdCI6MTU5MDY3NzE4MiwiZXhwIjoxNTkwNjc4OTgyfQ.KNE2ckvmVrx8myJfkAuNRc1zKfaSdsAAYR6dTCeWMh4
2020-05-28 10:03:08.885 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.035 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.035 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.036 [Verbose] [main] [WebViewController.swift:584] userContentController(_:didReceive:) > getExternalAuth called
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:09.067 [Verbose] [main] [WebViewController.swift:596] userContentController(_:didReceive:) > Success on getExternalAuth callback: nil
2020-05-28 10:03:11.675 [Verbose] [main] [HAAPI.swift:506] GetAndSendLocation(trigger:) > getAndSendLocation called via Manual
2020-05-28 10:03:11.676 [Verbose] [main] [OneShotLocationManager.swift:36] init(onLocation:) > LocationManager: Got location stored on manager <+33.83813477,-118.33911486> +/- 65.00m (speed -1.00 mps / course -1.00) @ 5/28/20, 10:03:08 AM Pacific Daylight Time
2020-05-28 10:03:11.683 [Info] [main] [HAAPI.swift:477] SubmitLocation(updateType:location:zone:) > Location update payload: ["altitude": 23.2812557220459, "vertical_accuracy": 10.0, "gps": [33.838134765625, -118.33911486011463], "battery": 99, "gps_accuracy": 65.0]
```

It appears that some token was 6405 seconds out of date and therefore it tried to auth and failed. I think this time I did a manual refresh rather than let it time out for 10 seconds. 

Seems to be something in WebViewController.

**Comment by TomBrien** [comment]
So looks like it is a timeout of the token as we were suspecting. Hopefully this will get fixed by Robbie next week

**Comment by mbaran5** [comment]
does it make sense to close this issue and re-open it as a ticket refresh issue with these logs? I can close the wonky GPS one as well since they seem to be related (I think)

**Comment by TomBrien** [comment]
Let’s just re name this one I think

**Comment by zacwest** [comment]
Talked with @mbaran5 on Discord - the problem is the 'Connection lost. Reconnecting…' banner at the bottom sticks around for a while before it retries. This should be a case we can detect and force to clear up.

**Comment by zacwest** [comment]
Looked at a few more logs, seems to always have multiple auth token refreshes requested by the webview at once, which can definitely cause an incorrect failed-login-attempt like:

1. webview asks for refreshed token
1. app starts refreshing with token A (request 1)
1. webview asks for refreshed token
1. app starts refreshing with token  A (request 2)
1. app gets new token from request 1
1. app triggers error using now-invalid token from request 2

in the end we're still logged in because request 1 successfully refreshed, but it likely caused some weird things in the middle.

**Comment by googanhiem** [comment]
The issue has reduced a lot since 0.112 (I guess the timeout helps with the auth reissues). It still happens occasionally, the event log doesn't have any further info, which makes it a bit hard to troubleshoot.

**Comment by googanhiem** [comment]
I'm still getting this issue just as frequently on 0.115 and 2020.6 (15). I don't have these issues while using the frontend in browser.
I've tested a number of changes in the last two months (switching to hassio, changing DNS provider, mDNS changes, reducing the frequency of iOS notifications) I've not been able to fix this one. 
At this point, the only thing I think could be causing this token mismatch is possibly having slow response times for auth on my system (rpi3, high latency connection, duckdns... takes about 500ms for server response on average). 
Are others still having this issue? Any changes remedy it?

**Comment by googanhiem** [comment]
Also @zacwest do you think #772 address this issue?

**Comment by zacwest** [comment]
I do not believe the WebSocket API will do anything to affect this issue. It would be useful to see logs from the time window of the connection issues you're experiencing, since that may be an indication of what is going on.

It is worth noting that the browser and the app will behave slightly different when it comes to authentication. The frontend uses a separate method to grab authentication tokens from the app rather than using its own, so it may be an issue in the frontend with how that works.

**Comment by zacwest** [comment]
I think I've tracked it down to an issue in the frontend: home-assistant/frontend#7062 -- if my hypothesis is correct, what's happening is that multiple things are requesting an authenticated call (connecting websocket, loading a camera, issuing a rest api call, etc.) cause all but the last one to be timing out silently. I do not believe this is an iOS issue at this point.

**Comment by googanhiem** [comment]
Looking at my logs this is probably the issue. I’m going to remove the offending elements and see if that makes a difference. Thanks for the insight

**Comment by googanhiem** [comment]
Yep, thanks @zacwest that was the fix. 

If anyone is having this issue, a temporary solution will be removing cards (from all views) for a camera, or element that triggers a separate https call (you can see this via chrome’s inspector if you’re unsure).

Again this is an issue with Home Assistants front end, not this iOS app, so this can probably be closed now.

**Comment by zacwest** [comment]
I believe this is now resolved as of HA 0.117.
